### PR TITLE
Make `get_default_canonical_id` public

### PR DESCRIPTION
**Pull request by Silic0nS0ldier**
When `repository_ctx.download` or `repository_ctx.download_and_extract` are not given an explicit `canonical_id` the default behaviour can lead to some counterintuitive results (e.g. URL changed but old asset restored from cache due to unchanged checksum).

This PR seeks to bring greater attention to `canonical_id` in these low level API (relative to `http_archive` which uses `get_default_canonical_id` by default).

URLs are usually the most appropriate `canonical_id` choice, so `get_default_canonical_id` has been added to the public API and sample usage added to documentation.

Related to #22652

## TODO

I need some pointers to make these happen.

- [x] Add `get_default_canonical_id` to the Bazel docs (I believe there are other `load`ed API that is missing from docs in one way or another).
- [x] Confirm examples render correctly in docs.

### CODE REVIEW

**Comment by Silic0nS0ldier** [comment]
Broken test _should_ be fixed, and I've got the documentation sorted.

**Comment by Silic0nS0ldier** [comment]
Maybe this time, since I apparently missed a default lockfile test. 🤞

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
+FROM ubuntu:24.10
```
Comment (fmeum) [inline .devcontainer/Dockerfile:1]: Could you split this into a separate PR if you think it could be useful to others?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java [hunk 2]
```diff
@@ -602,7 +602,16 @@ private StructImpl completeDownload(PendingDownload pendingDownload)
           "Downloads a file to the output path for the provided url and returns a struct"
               + " containing <code>success</code>, a flag which is <code>true</code> if the"
               + " download completed successfully, and if successful, a hash of the file"
-              + " with the fields <code>sha256</code> and <code>integrity</code>.",
+              + " with the fields <code>sha256</code> and <code>integrity</code>."
+              + "Setting an explicit <code>canonical_id</code> is highly recommended. e.g."
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java:606]: I find this recommendation a bit too broad: If this were the case, wouldn't we always want to use canonical ID? Instead, I think that we should recommend it in the case where the hash is user-specified rather than generated automatically.

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (Silic0nS0ldier) [reply .devcontainer/Dockerfile:1]: This actually got in by mistake while resolving a merge conflict. A dev container likely would be useful for contributors, but I'm not ready to make a push for it just yet (especially as building from `Dockerfile` is the slowest way to use dev containers).

I'll get this removed.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java (hunk 2 above)
Comment (Silic0nS0ldier) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java]: I've amended the recommendation to be specific to when `sha256` or `integrity` are user specified.

**Review by fmeum — APPROVED** [review]

**Review by meteorcloudy — APPROVED** [review]

**Comment by meteorcloudy** [comment]
@Silic0nS0ldier Can you please add an integration test to make sure this API would work as expected?
Maybe in https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/external_integration_test.sh

**Review by meteorcloudy — CHANGES_REQUESTED** [review]
Please add a simple integration test

**Comment by iancha1992** [comment]
Also, looks like there is a conflict with a file. Could you please take a look? Thanks!
cc: @meteorcloudy @Silic0nS0ldier @fmeum

**Comment by Silic0nS0ldier** [comment]
Conflict resolved. Working on integration test now.

**Comment by Silic0nS0ldier** [comment]
Integration test added, but I noticed `BAZEL_HTTP_RULES_URLS_AS_DEFAULT_CANONICAL_ID`. Do we want this behaviour outside of the builtin HTTP rules?

**Comment by meteorcloudy** [comment]
I guess we do want to keep it consistent? @fmeum WDYT?

**Comment by fmeum** [comment]
This is tough... The name suggests it only affects the `http_*` rules, but it's also surprising if the central knob doesn't control this for all similar rules. Maybe we interpret this as meaning "http_* like rules"?

**Comment by meteorcloudy** [comment]
It's not ideal, but I think it's fine with proper documentation?

**Comment by Silic0nS0ldier** [comment]
I'll make sure this is properly documented. I'm also going to move the examples over to `get_default_canonical_id`, having `.download` and `.download_and_extract` link to its documentation.

Side note, I noticed `download_remote_files` (a utility for downloading files which are overlaid into the repo) is used in the definition of `http_archive` but does not adhere to `BAZEL_HTTP_RULES_URLS_AS_DEFAULT_CANONICAL_ID`. This is in conflict with documented behaviour. Worse is that it is using the canonical ID of the 'main' file which seems like a bug (same canonical ID used for multiple files).

**Comment by fmeum** [comment]
It makes sense to change the behavior, although remote files are mostly used to power BCR use cases, where the URLs are not user-specified. It may even be more efficient to *not* use canonical IDs for them as many remote overlay files won't change between module versions, but their URLs will.

**Comment by Silic0nS0ldier** [comment]
Documentation has been updated.

I won't make any to `download_remote_files` in this PR (outside of scope).
